File: Striver/DP/Q50.java
package Striver.DP;

import java.util.Arrays;

public class Q50{
    // Minimum Cost to Cut the Stick

    // recursive solution - Time: O(2^n) Space: O(n)
    public static int minCost(int[] arr, int i, int j){
        if(i > j) return 0;
        int min = Integer.MAX_VALUE;
        for(int k = i; k <= j; k++){
            int temp = minCost(arr, i, k-1) + minCost(arr, k+1, j) + arr[j+1] - arr[i-1];
            min = Math.min(min, temp);
        }

        return min;
    }

    // memoization - Time: O(n^3) Space: O(n^2)
    public static int minCostMemo(int[] arr, int i, int j, int[][] dp){
        if(i > j) return 0;
        if(dp[i][j] != 0) return dp[i][j];
        int min = Integer.MAX_VALUE;
        for(int k = i; k <= j; k++){
            int temp = minCostMemo(arr, i, k-1, dp) + minCostMemo(arr, k+1, j, dp) + arr[j+1] - arr[i-1];
            min = Math.min(min, temp);
        }

        return dp[i][j] = min;
    }

    // tabulation - Time: O(n^3) Space: O(n^2)
    public static int minCostTab(int[] arr, int n){
        int c = arr.length;
        int[][] dp = new int[c+2][c+2];

        for(int i = c; i >= 1; i--){
            for(int j = 1; j <= c; j++){
                if(i > j) continue;
                int min = Integer.MAX_VALUE;
                for(int k = i; k <= j; k++){
                    int temp = dp[i][k-1] + dp[k+1][j] + arr[j+1] - arr[i-1];
                    min = Math.min(min, temp);
                }
                dp[i][j] = min;
            }
        }

        return dp[1][c];
    }

    public static void main(String[] args) {
        int[] arr = {1, 3, 4, 5, 8};
        int c = arr.length;
        int n = 12;
        int[] arr2 = new int[c+2];
        arr2[0] = 0;
        arr2[c+1] = n;
        for(int i = 1; i <= c; i++){
            arr2[i] = arr[i-1];
        }
        Arrays.sort(arr2);
        System.out.println(minCost(arr2, 1, c));
        System.out.println(minCostMemo(arr2, 1, c, new int[c+2][c+2]));
        System.out.println(minCostTab(arr2, n));
    }

}

File: Striver/DP/Q50.py
# minimum cost to cut the stick

# recursive solution
def minCost(cuts, i, j):
    if i > j:
        return 0
    mn = float('inf')
    for k in range(i, j+1):
        mn = min(mn, minCost(cuts, i, k-1) + minCost(cuts, k+1, j) + cuts[j+1] - cuts[i-1])
    return mn

# memoization
def minCostMemo(cuts, i, j, dp):
    if i > j:
        return 0
    if dp[i][j] != -1:
        return dp[i][j]
    mn = float('inf')
    for k in range(i, j+1):
        mn = min(mn, minCostMemo(cuts, i, k-1, dp) + minCostMemo(cuts, k+1, j, dp) + cuts[j+1] - cuts[i-1])
    dp[i][j] = mn
    return mn

# tabulation
def minCostTab(cuts, n):
    c = len(cuts)-2

    dp = [[0 for i in range(c+2)] for j in range(c+2)]
    for i in range(c, 0, -1):
        for j in range(1, c+1):
            if i > j:
                continue
            mn = float('inf')
            for k in range(i, j+1):
                mn = min(mn, dp[i][k-1] + dp[k+1][j] + cuts[j+1] - cuts[i-1])
            dp[i][j] = mn

    return dp[1][c]

arr = [1, 3, 4, 5, 8]
c = len(arr)
n = 12
cuts = [0] + arr + [n]
dp = [[-1 for i in range(c+2)] for j in range(c+2)]
cuts.sort()
print(minCost(cuts, 1, c))
print(minCostMemo(cuts, 1, c, dp))
print(minCostTab(cuts, n))